File: scripts/algorithms/C/Create Sorted Array through Instructions/Create Sorted Array through Instructions.cpp
// Runtime: 1154 ms (Top 36.85%) | Memory: 229 MB (Top 24.81%)
class Solution {
public:

    vector<int>seg;

    int query(int index,int low,int high,int l,int r)
    {
        if(low>=l && high<=r)
        {
            return seg[index];
        }
        if(high<l || low>r)
            return 0;

        int mid=(low+high)/2;
        int left=query(2*index+1,low,mid,l,r);
        int right=query(2*index+2,mid+1,high,l,r);
        return right+left;

    }

    void update(int index,int low,int high,int pos)
    {
        if(low==high)
        {
            seg[index]++;
            return ;
        }
        else
        {
            int mid=(low+high)/2;
            if(pos<=mid)
            {
                update(2*index+1,low,mid,pos);
            }
            else
            {
                update(2*index+2,mid+1,high,pos);
            }
            seg[index]=seg[2*index+1]+seg[2*index+2];
        }
    }

    int createSortedArray(vector<int>& instructions) {

        // using 100001 as 1e5 is the max element value
        seg.resize(4*100001,0);
        int mod=1e9+7;
        int ans=0;
        for(int i=0;i<instructions.size();i++)
        {
            int mi=query(0,0,100001,0,instructions[i]-1); //number of elements smaller than instruction[i]
            int mx=query(0,0,100001,instructions[i]+1,100001); // number of elements greater than instruction[i]
            update(0,0,100001,instructions[i]);
            ans=(ans+min(mi,mx))%mod;
        }
        return ans;

         /*TLE approch using STL
         // 50/65 cases passed
        int m=instructions.size();
        vector<int>a;
        int mod=1e9+7;
        int ans=0;
        int mxe=INT_MIN;
        unordered_map<int,int>um;
        for(auto x:instructions)
        {
            int xx=lower_bound(a.begin(),a.end(),x)-a.begin();
            int n=a.size();
            um[x]++;
            mxe=max(mxe,x);
            if(n==0)
            {
                a.push_back(x);

            }
            else
            {
                a.insert(a.begin()+xx,x);

                int mi=xx;
                int m=a.size();
                ans=(ans+min(mi,m-mi-um[x]))%mod;

            }
        }

        return ans;*/

    }
};

File: scripts/algorithms/C/Create Sorted Array through Instructions/Create Sorted Array through Instructions.js
// Runtime: 3484 ms (Top 12.50%) | Memory: 78 MB (Top 37.50%)
var createSortedArray = function(instructions) {
    let nums = [];
    let cost = 0;
    // define a map to track how many occurences of a number are in the array as we sort it
    let map = {};
    // classic binary search
    const bin = (arr, elem) => {
        let low = 0;
        let high = arr.length - 1;
        while (low <= high) {
            let mid = (high + low) >> 1; // fancy Math.foor(high + low / 2)
            if (arr[mid] > elem) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
    // loop through the each of the values
    for (let i of instructions) {
        // do a binary search to determine which index this value belongs at
        let index = bin(nums, i);
        // insert it into the array
        nums.splice(index, 0, i);
        // populate a map to track how many instances of this number we have in the arr
        if (!map.hasOwnProperty(i)) map[i] = 0;
        // need to remove any matching numbers from the equation, meaning if the left side is 1,2,3,3
        // and the current value is 3, we should be subtracting the existing 3's from the equation
        let left = index - map[i];
        // right side is the last value - index
        let right = (nums.length - 1) - index;
        // add the min of left / right to the cost
        cost += Math.min(left, right);
        // increment our number occurence map
        map[i]++;
    }

    const WEIRD_MOD_VAL = Math.pow(10, 9) + 7;
    // for some goofy reason when the number is too large, send the mod val
    return cost < 1842204555 ? cost : cost % WEIRD_MOD_VAL;
};